pymc_marketing/clv/distributions.py

```diff
@@ -460,3 +457,189 @@ def logp(value, r, alpha, s, beta, T):
             beta > 0,
             msg="r > 0, alpha > 0, s > 0, beta > 0",
         )
+
+
+class BetaGeoBetaBinomRV(RandomVariable):
+    name = "beta_geo_beta_binom"
+    ndim_supp = 1
+    ndims_params = [0, 0, 0, 0, 0]
+    dtype = "floatX"
+    _print_name = ("BetaGeoBetaBinom", "\\operatorname{BetaGeoBetaBinom}")
+
+    def make_node(self, rng, size, dtype, alpha, beta, gamma, delta, T):
+        alpha = pt.as_tensor_variable(alpha)
+        beta = pt.as_tensor_variable(beta)
+        gamma = pt.as_tensor_variable(gamma)
+        delta = pt.as_tensor_variable(delta)
+        T = pt.as_tensor_variable(T)
+
+        return super().make_node(rng, size, dtype, alpha, beta, gamma, delta, T)
+
+    def __call__(self, alpha, beta, gamma, delta, T, size=None, **kwargs):
+        return super().__call__(alpha, beta, gamma, delta, T, size=size, **kwargs)
+
+    @classmethod
+    def rng_fn(cls, rng, alpha, beta, gamma, delta, T, size) -> np.ndarray:
+        size = pm.distributions.shape_utils.to_tuple(size)
+
+        alpha = np.asarray(alpha)
+        beta = np.asarray(beta)
+        gamma = np.asarray(gamma)
+        delta = np.asarray(delta)
+        T = np.asarray(T)
+
+        if size == ():
+            size = np.broadcast_shapes(
+                alpha.shape, beta.shape, gamma.shape, delta.shape, T.shape
+            )
+
+        alpha = np.broadcast_to(alpha, size)
+        beta = np.broadcast_to(beta, size)
+        gamma = np.broadcast_to(gamma, size)
+        delta = np.broadcast_to(delta, size)
+        T = np.broadcast_to(T, size)
+
+        output = np.zeros(shape=size + (2,))
+
+        purchase_prob = rng.beta(a=alpha, b=beta, size=size)
+        churn_prob = rng.beta(a=delta, b=gamma, size=size)
+
+        def sim_data(purchase_prob, churn_prob, T):
+            t = 0
+            n = 0
+            active = True
+
+            while t <= T and active:
+                t += 1
+                active = rng.binomial(1, churn_prob)
+                purchase = rng.binomial(1, purchase_prob)
+                if active and purchase:
+                    n += 1
+
+            return np.array(
+                [
+                    t,
+                    n,
+                ],
+            )
```

Comment:
Is `t` and `n` supposed to be `t_x` and `x` from the paper? If so, it looks like `t` should be the time of the last observation and not the time of death (which is supposed to be unknown)?

Otherwise the draws from this RV don't match with the logp method

Reply:
Also I think we can avoid the while loop. Untested code:

```python
t_death = pt.minimum(rng.geometric(churn_p), T)   # May need an offset on the geometric
purchases = rng.binomial(t_death, purchase_prob)
x = np.sum(purchases)
t_x = T - np.argmax(purchases[::-1])
```

Reply:
@ColtAllen did you see this comment :arrow_up: ?